stellar-contract-env-host/src/host.rs

```diff
@@ -1549,30 +1549,131 @@ impl CheckedEnv for Host {
         Ok(self.to_host_obj(&sco)?.into())
     }
 
-    fn binary_copy_to_guest_mem(
+    fn binary_copy_to_linear_memory(
         &self,
         b: Object,
         i: RawVal,
         j: RawVal,
         l: RawVal,
     ) -> Result<RawVal, HostError> {
-        todo!()
+        //[offset in the host binary, pos in the linear memory, length of the slice to copy]
+        #[cfg(not(feature = "vm"))]
```

Comment:
Why do some of the host functions require the full VM? It feels like we're entering a mode where the host crate should just always be in VM mode, and we should remove the feature because it always needs to be on. cc @graydon @jonjove 

(This comment doesn't need to block this PR.)

Reply:
Yeah these ones do, since they need access to the linear memory, which is part of the VM. 
I am certainly in favor of removing the VM feature toggle.

Reply:
To me, having the "vm" feature be explicit has been helpful in making sure that the host can be meaningfully configured/built in its "native" mode (where guest calls host directly) -- I am certain we're not passing through any VM infrasturcture along the way because it's all compiled-out. So I'd _like_ to maintain it, but won't insist if it's getting too painful.

That said, in this case making the "vm" feature mandatory wouldn't make this (or _any_ of the linear-memory-accessing functions) make any more sense for the native-build (guest == host) case. Indeed, it's good that we got stopped here to consider it, because this is one of the cases where we're at risk of making the native build incoherent. These functions simply don't make sense as-written for native builds. They either need to have different signatures or else be avoided in native builds, be performed SDK-side.

Think about what the parameters to these functions mean: they're 32-bit numbers that represent portions of the guest memory. In a native build, guest memory is host memory. This causes two issues:

  1. That guest/host memory is probably 64 bit, not 32 bit.
  2. Even if you make it handle 64-bit correctly, it's not entirely clear we'll be able to have (or keep) a non-UB interface to converting from a u64 back to a pointer (under the forthcoming [strict provenance](https://github.com/rust-lang/rust/issues/95228) model in Rust, anyway).

I'm not 100% sure what the best way around this is right now.

The simplest thing -- and what I'd suggest trying "for now" -- would be to define it to work with arguments of multiple integer types -- u32, u63, or even u64-in-an-object -- and implement the non-VM case by converting the values back to a raw pointers (like just using `u64 as *mut u8`) and doing an unsafe memcpy with the resulting pointers. This will work in today's rust but may be harder to support in the future (or require more elaborate ceremony to tell the compiler "we know what we're doing" -- the strict provenance docs _say_ they're going to keep supporting this kind of conversion, somehow). In the VM case, just trap on anything non-u32.

There might be other more-awkward options to explore in the future, such as injecting a `base pointer` into the host when it's constructed and requiring all passed pointers to guest-to-binary / binary-to-guest functions be within the same provenance as the base pointer, or something, and then doing the guest-to-guest operations strictly SDK-side.

Reply:
@leighmcculloch I'm not totally sure this _can_ be done SDK-side, at least not with a lot of mode-specific logic (not sure how much of that you have already); does the SDK have the ability to retrieve a "true" host object pointer through which it could locate the interior of a Binary, rather than by handle?

Reply:
We could add functions to the `Host` to expose that I guess. The SDK can call fns added to Host that are not part of the `Env` trait.